Introduce prune-stdio

gkleen opened the pull request:
This pull request introduces `--keep-stdio` to `borg prune`.  
Instead of considering the other `--keep-*` options this mode outputs all archives (excluding checkpoints) on stdout and keeps exactly those archives returned on stdin.

Fixes #6266

infectormp commented:
It would be nice if you could add examples of use to the documentation

gkleen commented:
> It would be nice if you could add examples of use to the documentation

@infectormp: done

ThomasWaldmann commented:
https://borgbackup.readthedocs.io/en/1.2.0rc1/usage/create.html

See `--paths-from-stdin` and `--paths-from-command` there for existing options that are somewhat similar.

Maybe name your option in a similar way if there is no reason to do it differently.

ThomasWaldmann commented on requirements.d/docs.txt:
```diff
@@ -1,2 +1,2 @@
-sphinx
+sphinx>=1.8.0
```
?

gkleen replied:
Commited by mistake; will remove...

ThomasWaldmann commented on src/borg/archiver.py:
```diff
@@ -4424,6 +4475,14 @@ def define_borg_mount(parser):
+        ``borg list``) and expect a json list of archives to keep on stdin.
+        Each element of the list is expected to be an object containing at least
+        one of the keys ``id`` or ``barchive``, as in the output.
+        Only archives given on stdin will be kept.
+        Empty stdin will be treated like ``[]``.
```
I see some danger of  people accidentally killing all archives because of some empty input...

gkleen replied:
True; will change...

ThomasWaldmann commented on src/borg/archiver.py:
```diff
@@ -1517,6 +1521,53 @@ def do_prune(self, args, repository, manifest, key):
+                output_data.append(formatter.get_item_data(archive))
+            json_print(basic_json_data(manifest, extra={'archives': output_data}))
+
+            def hex_to_bin(d):
+                bin_id = None
+                try:
+                    bin_id = unhexlify(d)
+                except:
+                    raise ValueError('Invalid value, must be 64 hex digits') from None
+                if len(bin_id) != 32:
+                    raise ValueError('Invalid value, must be 64 hex digits')
+                return bin_id
```
also `except:` should be more specific usually.

ThomasWaldmann replied:
why `from None` in one case, but not in the other?

gkleen replied:
I looked around a bit and shamelessly stole that code from `do_config` in `archiver.py` (the same file).
Weird exception handling and all.

I'll factor both usage sites out to `borg.helpers` and improve exception handling along the way?

ThomasWaldmann replied:
sounds good. wondering that we do not already have that there?

gkleen replied:
see below

ThomasWaldmann commented on src/borg/archiver.py:
```diff
@@ -1517,6 +1521,53 @@ def do_prune(self, args, repository, manifest, key):
+            try:
+                input = sys.stdin.read()
+                input_data = []
+                if input:
+                    input_data = json.loads(input)
```
i think we must require the input to be always valid json (even if it is just `[]`) and reject everything else, including the empty (`''`) input.

gkleen replied:
See abov; will change...

ThomasWaldmann reviewed (commented):
thanks for the PR, some first feedback.

gkleen commented on docs/usage/prune.rst:
```diff
@@ -34,6 +34,62 @@ first so you will see what it would do without it actually doing anything.
+    Keeping archive (rule: stdio #2):        archive_1                            Sun, 2022-02-13 11:44:37 [4ed9fa060f7a37bde7408c3cc1c32f1371afaf49c478379c46ca13d3e0b86db4]
+    Pruning archive (2/2):                   archive_1.checkpoint                 Sun, 2022-02-13 11:44:17 [92662ffd369b20c00168bf9eff821a6c00efe6e622cf1e7bce13e6f20fda5643]
+
+It is up to the user to generate the stdin from borgs output.
+Usually `borg prune --keep-stdio` would be run as a subprocess from a custom script.
```
The third exchange only takes place because `--list` is used for clarity.  
In a script I would expect stdin and stdout to be used for interaction with the script and stderr passed through to the user.  
Does that make sense?

gkleen replied:
There is now also a mode that uses a delimiter character instead of json

gkleen commented on src/borg/archiver.py:
```diff
@@ -4424,6 +4475,14 @@ def define_borg_mount(parser):
         keep the last N archives under the assumption that you do not create more than one
         backup archive in the same second).
 
+        When using ``--keep-stdio`` borg will produce a listing of all
+        non-checkpoint archives on stdout (similiar to the behaviour of
+        ``borg list``) and expect a json list of archives to keep on stdin.
```
The problem with that is locking.  
In theory one could use `borg with-lock` but then every "inner" borg command would need `--bypass-lock` or similiar; that seems very janky to me.

ThomasWaldmann replied:
another idea could be to refactor this internally into 2 commands and then call common code by such a high-level command.

so, make borg list be able to do what you want (output json in a suitable way).
make borg prune be able to do what you want (accept json input).
call do_list / do_prune (or so) by your code (could be a new high-level command, not inside prune).

gkleen replied:
I experimented a bit and found a way to reuse quite a bit more code by switching from an option `--keep-stdio` to a new subcommand `prune-stdio`.

gkleen commented on src/borg/archiver.py:
```diff
@@ -4460,6 +4519,8 @@ def define_borg_mount(parser):
                                help='number of monthly archives to keep')
         subparser.add_argument('-y', '--keep-yearly', dest='yearly', type=int, default=0,
                                help='number of yearly archives to keep')
+        subparser.add_argument('--keep-stdio', dest='filter_stdio', action='store_true',
```
`keep_stdio` it is; will change...

gkleen commented:
> https://borgbackup.readthedocs.io/en/1.2.0rc1/usage/create.html
> 
> See `--paths-from-stdin` and `--paths-from-command` there for existing options that are somewhat similar.
> 
> Maybe name your option in a similar way if there is no reason to do it differently.

One large difference is that this option also produces some defined output on `stdout` first.  
I'd therefore like to keep `stdio` or both `stdin` and `stdout` in the name.

Additionally I believe that keeping the option name close to the other `--keep-*`options provides some value.

I currently don't have a better suggestion than `--keep-stdio` or `--keep-via-stdio` (with a weak preference for `--keep-stdio`). Anybody else?

gkleen commented:
There seems to have been quite a few changes since the last `python setup.py build_usage`/`… build_man`.  
Should I try to manually remove the patch hunks that don't pertain to this pull request directly (I'd rather not...)?

ThomasWaldmann commented:
Please don't commit build_usage / build_man results (I do that at release time).

Keeps the PRs smaller / better reviewable.

gkleen commented:
> Please don't commit build_usage / build_man results (I do that at release time).
> 
> Keeps the PRs smaller / better reviewable.

"None at all" is an easy fix; done

gkleen commented:
@ThomasWaldmann is this good to merge in principle?  
I assume it won't make it for the first 1.2 release?

fantasya-pbem commented:
Just a question on

> This pull request [...] keeps exactly those archives returned on stdin.

- why is it implemented as "keep what is returned" and not "prune what is returned"?

gkleen commented:
> - why is it implemented as "keep what is returned" and not "prune what is returned"?

By analogy to the other `--keep-*` options and the `--paths-from-*` options on `borg create`

gkleen commented:
I don't have much bandwidth to retool this for borg 2.0 – going to investigate a solution involving `borg with-lock`, instead...
